aioredis/cluster/cluster.py

```diff
@@ -0,0 +1,615 @@
+
+        raise RedisClusterError(
+            "No cluster info could be loaded from any host")
+
+    @asyncio.coroutine
+    def initialize(self):
+        logger.info('Initializing cluster...')
+        self._moved_count = 0
+        yield from self.fetch_cluster_info()
+        logger.info('Initialized cluster.\n{}'.format(self._cluster_manager))
+
+    @asyncio.coroutine
+    def clear(self):
+        pass  # All connections are created on demand and destroyed afterwards.
+
+    @asyncio.coroutine
+    def create_connection(self, address):
+        conn = yield from create_redis(
+            address, db=self._db, encoding=self._encoding,
+            password=self._password, loop=self._loop, )
+        return conn
+
+    @asyncio.coroutine
+    def _execute_node(self, address, command, *args, **kwargs):
+        """Execute redis command and returns Future waiting for the answer.
+
+        :param command str
+        :param pool obj
+        Raises:
+        * TypeError if any of args can not be encoded as bytes.
+        * ReplyError on redis '-ERR' responses.
+        * ProtocolError when response can not be decoded meaning connection
+          is broken.
+        """
+        cmd = decode(command).lower()
+        to_close = []
+        try:
+            conn = yield from self.create_connection(address)
+            to_close.append(conn)
+            return (yield from getattr(conn, cmd)(*args, **kwargs))
+        except ReplyError as err:
+            address = parse_moved_response_error(err)
+            if address is None:
+                raise
+            logger.debug('Got MOVED command: {}'.format(err))
+            self._moved_count += 1
+            if self._moved_count >= self.MAX_MOVED_COUNT:
+                yield from self.initialize()
+                node = self.get_node(command, *args, **kwargs)
+                address = node.address
+            conn = yield from self.create_connection(address)
+            to_close.append(conn)
+            return (yield from getattr(conn, cmd)(*args, **kwargs))
+        finally:
+            for conn in to_close:
+                conn.close()
+                yield from conn.wait_closed()
+
+    @asyncio.coroutine
+    def _execute_nodes(self, command, *args, **kwargs):
+        """
+        Execute redis command for all nodes and returns
+        Future waiting for the answer.
+
+        :param command str
+        Raises:
+        * TypeError if any of args can not be encoded as bytes.
+        * ReplyError on redis '-ERR' responses.
+        * ProtocolError when response can not be decoded meaning connection
+          is broken.
+        """
+        return (yield from asyncio.gather(
+            *[self._execute_node(node.address, command, *args, **kwargs)
+              for node in self._cluster_manager.masters],
+            loop=self._loop))
+
+    @asyncio.coroutine
+    def execute(self, command, *args, many=False, **kwargs):
```

Comment:
The problem in the cluster case is that I cannot know to which server I should direct unknown commands. Currently, all commands with a key are sent to the server responsible for the key. All other commands are simply sent to all masters, which makes sense for commands like FLUSHALL, but doesn't make sense for many others.

Possibly this is also a problem in your pool refactoring, because it seems that now the pool class must figure out the right server based on a command name and arguments.